### ISSUE TITLE
Suggestion for improvement of the BulkWrite API 

### ISSUE THREAD

**Issue by petr-buchin** [mongodb/mongo-php-library#209]
I'm writing a new MongoDB ODM for PHP7, which, like Doctrine, will deal with bulk writes. 

I'd like to build ODM on top of MongoDB PHP Library, but here is what I found:
Driver's BulkWrite class has a beautiful API, which incapsulates functionality of building bulk write in methods like `insert()`, `update()`, `delete()` etc. 
On the other hand, the implementation of BulkWrite in the PHP Lib contains an ugly API, which is built on top of the driver's beautiful API: 
- BulkWrite accepts multidimensional array of operations, which looks awkward compared with the driver's solution with methods;
- PHP LIb API is NOT encapsulates driver's BulkWrite functionality, because if you want to build bulk write using php library, you must use constants from driver's BulkWrite class to specify, what operation should be done (update, insert or delete). As another solution, you must use string literals to specify operation. But this solution is prone to typos and looks even more awkward. So, there is a choice from two bad solutions: using constants from driver class, which functionality must be encapsulated in php lib, or using string literals.

I understand, why bulk write API in  PHP Lib is built this way: that's because Driver's BulkWrite class accepts options in constructor, and to use it's methods like `insert()` or `delete()`, we must specify all of the options first, but you want to have ability to validate/add/change/process some options before passing them to driver's BulkWrite. Also, maybe you want to create driver's BulkWrite lazily - when someone calls `execute()` on lib's BulkWrite (but that's just an object, and it's constructor does not open connection to Mongo server, so I think this is ok to create driver's bulk write in lib's BulkWrite constructor and decorate it's methods - this also gives ability to validate options and operations using driver's BulkWrite object).

Generally, if someone wants to build some kind of query builder on top of the PHP Lib, he must build cool bulk write API on top of ugly PHP Lib API, which is built on top of cool driver's API. From my opinion -  that's not cool. 

**Suggestion:**
I suggest to change `MongoDB\Operation\BulkWrite` constructor to NOT accept $operations argument.
Than I'd like to add methods `insert()`, `update()`, `delete()`, `flush()` and `setOptions()` to make building bulk write easy and cool. Method setOptions is not necessary if the $options argument in constructor will be required - if so, we can create driver's BulkWrite in constructor: than methods like insert() and others should just decorate methods from driver implementation. 

```
$bulk = new BulkWrite($options);
$bulk
    ->insert(['foo' => 'bar'])
    ->update(['foo' => 'bar'], ['$set' => ['Bar' => 'Baz']])
    ->delete(['foo' => 'Baz'])
    ->flush();
```

The next thing that needs to be done, is to change the API of the Collection class: i propose to add method `createBulkWrite(array $options = [])`, which will return `MongoDB\Operation\BulkWrite` instance:

```
// flush() method internally calls BulkWrite::execute();
$collection
    ->createBulkWrite()
    ->insert($insert)
    ->update($filter, $update)
    ->setOptions($options)
    ->flush();
```

What are you think, guys? 
If you like this idea, I would like to implement it and create a pull request.
Thanks in advance for reading a lot of text :)

**Comment by jmikola** [comment]
The current API is admittedly clunky, but was needed to implement our cross-driver [CRUD specification](https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst), which defines a `Collection::bulkWrite()` method.

The alternative to the current array structure would have been a number of model classes (one per operation), and then have `Collection::bulkWrite()` accept an array of those. I don't recall the exact reasons we went with the array implementation (perhaps an efficiency argument was made for arrays), but I do think the object route would be much cleaner. Those objects could then validate arguments for each operation, as the single-write Operation classes currently do. This would avoid the mess that is [`MongoDB\Operation\BulkWrite::__construct()`](https://github.com/mongodb/mongo-php-library/blob/cd9394c24aaee46864963cdff63c8b1c89c35bde/src/Operation/BulkWrite.php#L73).

The fluent API you're suggesting may conflict with the need to capture the return value of [`BulkWrite::insert()`](http://php.net/manual/en/mongodb-driver-bulkwrite.insert.php), which is the only time we can obtain the generated ID of an inserted document to populate the WriteResult object returned after execution.

> I suggest to change MongoDB\Operation\BulkWrite constructor to NOT accept $operations argument. Than I'd like to add methods insert(), update(), delete(), flush() and setOptions() to make building bulk write easy and cool.

Changing `Collection::bulkWrite()` to no longer accept an array of operations is not an option, for two reasons:

 * Implementing the CRUD spec is non-negotiable, as we need to ensure consistency across various language drivers
 * Changing the signature now would be a BC break, which would require a bump to 2.0

I do like the idea of either having a builder object, or simply taking an array of model classes, which validate their arguments upon construction. Actual execution and creation of the driver's BulkWrite object may need to stay in the `MongoDB\Operation\BulkWrite` class in order to capture generated IDs, though.

**Comment by petr-buchin** [comment]
Thank you for the detailed answer.
I'll try to implement solution with the array of models, and maybe I will create some kind of buider on top of this solution. I will not change the signature of the existing methods, but accepting an array of models instead of array in format [$operation => $args], will, of course, break BC indirectly.
Do I need to implement the solution, which can accept array in the old format AND array of models, or I can break BC, implementing the solution, which accept only array of models?

**Comment by lindelius** [comment]
I agree, as well, that the current array-based solution isn't ideal. However, I'd rather see a `BulkBuilder` aproach than keeping the array aproach and populating it with objects instead of associative arrays. Creating several new classes just seems overkill, to be honest, and the following example looks much better in my eyes.

```php
$bulk = new BulkBuilder();
$bulk->deleteOne($filter);
$bulk->deleteMany($filter);
$bulk->insertOne($document);
$bulk->replaceOne($filter, $replacement, $options);
$bulk->updateOne($filter, $update, $options);
$bulk->updateMany($filter, $update, $options);

$result = $collection->bulkWrite($bulk, $options);
```

**Comment by petr-buchin** [comment]
@lindelius , that's what I proposed, with the difference, that I proposed to add this method to `MongoDB\Operation\BulkWrite` class. But yes, that's silly, and moving this methods to the builder object is a better idea. 
But as @jmikola said, changing the signature of the `Collection::bulkWrite` method is just not an option.

Anyway, I think we can accomplish this API by doing so: 
```php
$bulk = new BulkBuilder();

$bulk->deleteOne($filter);
$bulk->deleteMany($filter);
$bulk->insertOne($document);
$bulk->replaceOne($filter, $replacement, $options);
$bulk->updateOne($filter, $update, $options);
$bulk->updateMany($filter, $update, $options);

$result = $collection->bulkWrite($bulk->getOperations(), $options);
```

BulkBuilder methods will create corresponding operations, and save them in its property, and than we can get this operations in format, suitable for `MongoDB\Operation\BulkWrite::__construct()` and `MongoDB\Collection::bulkWrite()`. This approach is not breaking backward compatibility and gives the API you are talking about.

The another problem is that options must be validated somewhere, and the current approach with copy-pasting a lot of if's from operation classes to the BulkBuilder methods or to objects, mentioned by @jmikola, is not looking too good. I'm thinking about implementing something like Symfony's OptionsResolver in order to reduce this code duplication. But the main point is that options must be validated somewhere. Objects, mentioned by @jmikola could do this, either by using some kind of OptionsResolver or just using if's.

**Comment by lindelius** [comment]
Having a `BulkBuilder` doesn't necessarily change the signature of the `bulkWrite()` method. And regarding the validation, there wouldn't really be that much duplication going on considering the bulk operations doesn't support all the various options that you can pass to the single-write classes. Most of the validation in the `BulkWrite` class is for validating the structure and operation types in the `$operations` array passed to the constructor, which would obviously have to remain as it is to keep BC.

**Comment by petr-buchin** [comment]
`MongoDB\Collection::bulkWrite()` method is just a shortcut for constructing and executing `MongoDB\Operations\BulkWrite` instance, so $operations argument should be the same for both, `MongoDB\Collection::bulkWrite()` and `MongoDB\Operations\BulkWrite::__construct()`.
Form @jmikola answer, we cannot change the signature of `MongoDB\Collection::bulkWrite()` is not an option at all, since collection class API **must** implement CRUD spec, which is common for all language drivers. An due to CRUD spec, collection's method `bulkWrite()` **must** accept a list of operations. The question is only, should "operation" be an array, or an object, i.e should `bulkWrite()` accept a list of associative arrays, or a list of objects.
From my opinion - a list of associative arrays is hard to read and write, and looks ugly.

Yes, `BulkBuilder` will not break BC and will not change `MongoDB\Collection::bulkWrite()` or  `MongoDB\Operations\BulkWrite::__construct()` signatures. It can help to build array of the operations, which should be passed to that methods. 
The only difference from your example is that we will not pass `BulkBuilder` instance itself to `MongoDB\Collection::bulkWrite()`. Instead, we would pass an array of operations, which were built by `BulkBuilder`.

**Your example**:
```php
$result = $collection->bulkWrite($bulk, $options);
```

What actually can be implemented:
```php
$result = $collection->bulkWrite($bulk->getOperations(), $options);
```

**Comment by lindelius** [comment]
If the first argument of the `BulkWrite` constructor absolutely has to be an array, then yes. But then the `BulkBuilder` approach wouldn't really help much in cleaning up the code as the current validation in the `BulkWrite` class would have to stay in place. The only benefit would be the abstraction of the creation of each operation's associative array for the bulk array. In this case, I'd opt for the "array of bulk operation classes" aproach, since that would actually clean up a lot of the messy validation going on in the `BulkWrite` class.

**Comment by petr-buchin** [comment]
> But then the BulkBuilder approach wouldn't really help much in cleaning up the code
Yes, but the goal of the `BulkBuilder` is to create a nice API for **mongo-php-library** users.
Cleaning up internal realization of this lib is another issue that should be done.
I'll try to accomplish both of this goals: nice API for users and a clean realization.

**Comment by jmikola** [comment]
Apologies for the lengthy reply, but here are my replies and observations while reading the thread from the post after my last.

> @petr-buchin: I will not change the signature of the existing methods, but accepting an array of models instead of array in format [$operation => $args], will, of course, break BC indirectly. Do I need to implement the solution, which can accept array in the "old" (current) format AND array of models, or I can break BC, implementing the solution, which accepts only array of models?

I don't think there's any need to break BC. If we went the route of creating model classes (each of which did their own argument validation), I expect we'd refactor the `MongoDB\Operation\BulkWrite` constructor to convert the current associative array into an array of those models. If we went the route of a builder class, it could simply have a method to export its state in the current associative array format.

In either case, the `Collection::bulkWrite()` signature can be relaxed to accept the new alternative.

On the node of model classes, [`MongoDB\Operation\CreateIndexes`](https://github.com/mongodb/mongo-php-library/blob/master/src/Operation/CreateIndexes.php) is an example of converting an input array into objects responsible for validation; however, this uses an internal IndexInput class, which I don't plan to publicize.

> @petr-buchin: I think the best solution is to support the both, an array in old (current) format and an array of models, as an argument $operations, but to trigger E_USER_DEPRECATED when receiving array in old format.

This sounds a bit premature. If the new style takes off, I could see doing this in a later minor release leading up to a bump to 2.0, though.

> I proposed to add this methods to `MongoDB\Operation\BulkWrite` class

For consistency with other Operation classes, I'd like to keep the class immutable. Currently, we construct it once and execute it via a single method.

> @petr-buchin: But as @jmikola said, changing the signature of the MongoDB\Collection::bulkWrite() method is just not an option.
>
> ...
>
> Due to @jmikola answer, we cannot change the signature of MongoDB\Collection::bulkWrite() at all, since collection class API must implement CRUD spec, which is common for all language drivers. And due to the CRUD spec, collection's method bulkWrite() must accept a list of operations.

Just to clarify: The signature can be changed without breaking BC. Removing the `array` type hint to allow the current array argument or a BulkWriteBuilder instance would be acceptable. The purpose of our specifications is to ensure consistent APIs exist across drivers, but it is not their intention to limit enhancement.

> @petr-buchin: But the main point is that options must be validated somewhere. Objects, mentioned by @jmikola, could do this, either by using some kind of OptionsResolver or just using if's.

I did consider using Symfony's OptionsResolver when first creating these operation classes, but I was concerned about overhead. This is certainly something I can consider down the road as internal refactoring if we can allocate time to run benchmarks.

> @lindelius: But, then the BulkBuilder approach wouldn't really help much in cleaning up the code as the current validation in the BulkWrite class would have to stay in place. The only benefit would be the abstraction of the creation of each operation's associative array for the bulk array. In this case, I'd rather see the "array of bulk operation classes" aproach, since that would actually clean up a lot of the messy validation going on in the BulkWrite class.

> @petr-buchin: I'm thinking about BulkBuilder, which will build "array of bulk operation classes".

This was the conclusion I was reaching while reading through the thread. Creating internal classes, like IndexInput for CreateIndexes, for the various write operations would let us consolidate input validation. The current `switch` in `MongoDB\Operation\BulkWrite::__construct()` could then be simplified to construct the appropriate class from the array. For instance, `['insertOne' => ['x' => 1]]` would translate to `new InsertOneInput(['x' => 1])`, which validates that the first argument is a document (i.e. array or object).

Similarly, each method on a builder object could forward its arguments to the appropriate input class.

Lastly, we'd need to convert this array of input objects into a `MongoDB\Driver\BulkWrite` suitable for execution. Collection of generated IDs would still need to be done in `MongoDB\Operation\BulkWrite` to populate the [`MongoDB\BulkWriteResult`](https://github.com/mongodb/mongo-php-library/blob/master/src/BulkWriteResult.php) object. I expect it'd be most efficient to pass a `MongoDB\Driver\BulkWrite` into each input object for modification, instead of having another series of `instanceof` checks in the operation class. The input class method that modifies `MongoDB\Driver\BulkWrite` should forward the return value from the corresponding `MongoDB\Driver\BulkWrite` method, which would allow the operation class to capture the generated ID from InsertOneInput.

**Comment by lindelius** [comment]
Sounds good. And just to clarify, I think a builder object would be great as long as we can pass it straight to the `BulkWrite` constructor without also having to manually call a method on it to get the operations array, e.g. `$collection->bulkWrite($bulk->getOperations(), $options)`. Since this isn't an issue, I think a builder object is the way to go.

A suggestion would be to go with the following structure for the bulk builder related files

```
src/
    Bulk/
        BulkBuilder.php
        DeleteManyInput.php
        DeleteOneInput.php
        InsertOneInput.php
        ReplaceOneInput.php
        UpdateManyInput.php
        UpdateOneInput.php
```

The bulk builder would then be used as the earlier suggestion:

```php
$bulk = new \MongoDB\Bulk\BulkBuilder();
$bulk->updateOne($filter, $update, $options);

$result = $collection->bulkWrite($bulk, $options);
```

The `BulkBuilder` would internally, as per your suggestions, populate an internal array with the `MongoDB\Bulk\xxxInput` classes via its public methods (e.g. `updateOne($filter, $update, $options)`).

The `BulkWrite`constructor would accept both the current associative array format and `BulkBuilder` objects. Associative arrays would be converted to an array of `MongoDB\Bulk\xxxInput` classes to make use of their internal validation (to keep the constructor as clean as possible, this functionality could be separated into its own method).

**Comment by lindelius** [comment]
@petr-buchin Did you start working on this?

**Comment by petr-buchin** [comment]
@lindelius , yes, I almost finished, but I didn't have time to write the tests yet.

**Comment by petr-buchin** [comment]
@jmikola, I made pull-request, but when working on this issue, I realised that current architecture of the write operations also should be improved. Input classes are partial copies of the corresponding Operation classes, i.e. `InsertOneInput` validates options the same way `MongoDB\Operations\InsertOne` does, the only difference is that `MongoDB\Operations\InsertOne` also accepts database and collection names and options, which should not be accepted by bulk input class, since they should be set on the `BulkWrite` itself (not on every single operation, that is added to the bulk).
I think in future we could change the architecture in such way: we have Input classes, which can validate their arguments and add themselves to the bulk, and Collection methods can validate bulk options, like `bypassDocumentValidation`. Collection methods also can create bulk (for write operations) and pass it to the corresponding input classes. 
At the moment I try to implement such logic in my own mongodb lib.
